envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py

```diff
@@ -105,29 +170,214 @@ def test_shutdown_closes_console_pipes(self):
-        with temp_filename("captured_stdout.txt") as captured_stdout:
-            with redirect_stdout_to(captured_stdout):
-                kernel = InternalIPKernel()
-                kernel.init_ipkernel(gui_backend=None)
-                kernel.shutdown()
+    def test_raw_print_err_logged(self):
+        test_message = "your hovercraft is full of eels"
+
+        # Outside the context of the kernel, raw_print_err prints as expected.
+        captured = []
+        with capture_stream(sys.__stderr__, callback=captured.append):
+            IPython.utils.io.raw_print_err(test_message)
+        self.assertEqual(len(captured), 1)
+        captured_stdout = captured[0].decode("utf-8")
+        self.assertIn(test_message, captured_stdout)
+
+        # Again, with a running kernel.
+        kernel = InternalIPKernel()
+        kernel.init_ipkernel(gui_backend=None)
+        captured = []
+        log_catcher = catch_logs(
+            "envisage.plugins.ipython_kernel", level=logging.WARNING)
+
+        try:
+            with log_catcher as log_records:
+                with capture_stream(sys.__stderr__, callback=captured.append):
+                    IPython.utils.io.raw_print_err(test_message)
+        finally:
+            kernel.shutdown()
+
+        # Check captured output (which should not contain the test message)
+        # and the INFO-level log records (which should).
+        messages = [
+            record.getMessage() for record in log_records
+            if record.levelno == logging.WARNING
+        ]
+        test_message_found = any(
+            test_message in message for message in messages)
+        self.assertTrue(
+            test_message_found,
+            msg="raw_print_err messages were not found in the logs"
+        )
+        self.assertEqual(len(captured), 1)
+        captured_stdout = captured[0].decode("utf-8")
+        self.assertNotIn(test_message, captured_stdout)
 
-            with open(captured_stdout, "rb") as f:
-                stdout = f.read().decode("utf-8")
+        # After shutdown, raw_print_err again prints to __stderr__.
+        captured = []
+        with capture_stream(sys.__stderr__, callback=captured.append):
+            IPython.utils.io.raw_print_err(test_message)
+        self.assertEqual(len(captured), 1)
+        captured_stdout = captured[0].decode("utf-8")
+        self.assertIn(test_message, captured_stdout)
 
-        self.assertNotIn("Ctrl-C will not work", stdout)
+    # Helper functions.
+
+    def objects_of_type(self, type):
+        """
+        Find and return a list of all currently tracked instances of the
+        given type.
+        """
+        return [
+            obj for obj in gc.get_objects()
+            if isinstance(obj, type)
+        ]
+
+    def create_and_destroy_kernel(self):
+        kernel = InternalIPKernel()
+        kernel.init_ipkernel(gui_backend=None)
+        kernel.new_qt_console()
+        kernel.new_qt_console()
```

Comment:
why do we need two calls to `new_qt_console` when creating a kernel?

Reply:
Is this just a check to ensure that we *can* open multiple qt consoles when the kernel is running?

Reply:
Yep. Just making sure that nothing goes wrong even when our kernel has multiple consoles attached. That's presumably a harder case than just a plain kernel with no consoles.